release-24.1: sql: avoid unnecessary singleflight requests for role membership cache

rafiss opened the pull request:
Backport 6/7 commits from #135852 and 1/1 commits from #136418.

/cc @cockroachdb/release

Release justification: major performance improvement; linked to a customer escalation

---

See individual commits for refactors that make the enhancement easier.

---

Since we use a query that scans the whole role_members table to
populate the membership cache, it is wasteful to launch multiple
singleflight requests for different users. This patch changes
the cache refresh logic of the singleflight so that it always populates
the entire cache, which means there will only ever be one in-flight
query per each node.

fixes https://github.com/cockroachdb/cockroach/issues/135931
Release note (performance improvement): Improved the internal caching
logic for role membership information. This reduces the latency impact
of commands such as `DROP ROLE`, `CREATE ROLE`, and `GRANT role TO user`,
which cause the role membership cache to be invalidated.

rafiss committed:
sql: remove sql.auth.resolve_membership_single_scan.enabled setting

This setting was added only as an escape hatch, and we've never needed
to recommend using it. Since the single scan implementation is simpler
anyway, removing the old code will make other refactors to this code
easier.

Release note (ops change): Removed the
sql.auth.resolve_membership_single_scan.enabled cluster setting. This
was added out of precaution in case it was necessary to revert back to
the old behavior for looking up role memberships, but this escape hatch
has never been needed in practice since this was added in v23.1.

rafiss committed:
sql: use a different query to load role memberships

This new query ensures that users who have no parent roles appear in the
query results. This will support a refactor that allows us to populate
the entire cache all at once, rather than doing extra work for each
user.

Release note: None

rafiss committed:
sql: move membership cache population into singleflight

This avoids the behavior of each goroutine doing the work of populating
the cache, which was entirely wasteful. It also supports a future
refactor to populate the whole cache all at once.

Release note: None

rafiss committed:
sql: use a cheaper role existence check for membership cache

Now that we return a row for every user that exists, regardless of
whether the user has any parent roles, we don't need to make a separate
query for role existence.

Release note: None

rafiss committed:
sql: add a function that resolves all role memberships

The new function returns a flattened map of the role hierarchy for each
role that exists.

This is a refactor that will make populating the membership cache all at
once easier.

Release note: None

rafiss committed:
sql: avoid unnecessary singleflight requests for role membership cache

Since we use a query that scans the whole role_members table to
populate the membership cache, it is wasteful to launch multiple
singleflight requests for different users. This patch changes
the cache refresh logic of the singleflight so that it always populates
the entire cache, which means there will only ever be one in-flight
query per each node.

This also requires us to invalidate the cache during the RESTORE job, as
that job writes directly into the system.users and system.role_members
tables. invalidating the cache will ensure that all nodes will refresh
their cache to see the new users.

Release note (performance improvement): Improved the internal caching
logic for role membership information. This reduces the latency impact
of commands such as `DROP ROLE`, `CREATE ROLE`, and `GRANT role TO user`,
which cause the role membership cache to be invalidated.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136121)
<!-- Reviewable:end -->

rafiss committed:
sql: use read timestamp from cache to read role-related system tables

To populate role-related virtual tables like pg_roles, we need to make
sure the timestamp we use to read the underlying system table is not
newer than the timestamp that was used to populate the role membership
cache. It is possible for these timestamps to differ because we use a
leased descriptor to check the validity of the membership cache.

This resolves an unreleased bug where a "user does not exist" error
could occur while scanning pg_roles.

Release note: None

rafiss committed:
logictest: deflake as_of test

We previously adjusted the expected errors in
4728949f96da35fd64b9818927d29c1e744e1ab7. There is another assertion
later in the test that needs the same update.

Release note: None

fqazi reviewed (changes requested):
> Reviewed 5 of 5 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136121)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @rafiss)

___
*[`pkg/sql/authorization.go` line 730 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136121#-ODDq3w2CRcezJkRUdCz:-ODDq3w2CRcezJkRUdD-:b-qdu8sk) ([raw file](https://github.com/cockroachdb/cockroach/blob/968af659bd208af296b262994c2c2a196b70ea58/pkg/sql/authorization.go#L730)):*
> ```Go
> 		ret[username.AdminRoleName()] = true
> 	}
> <<<<<<< HEAD
> ```

There's a merge issue in this one


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

There was a broken commit, but the final result looks good

> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 5 of 5 files at r6, 7 of 7 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136121#-:-ODH3s0d5B7tHAdXX3N4:b-nsvm65)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kev-cao and @rafiss)



<!-- Sent from Reviewable.io -->
